File: adventofcode/2022/day11/day11.rb
# frozen_string_literal: true

def read_input
  ARGF.read
end

class Monkey
  attr_reader :total_inspected, :test
  attr_accessor :index, :items, :operation, :test_target

  def initialize(lines)
    @index = lines[0].split(":").first.split.last.to_i
    @items = lines[1].split(":").last.split(",").map(&:to_i)

    @op = lines[2].split(' = ').last
    @operation = Proc.new {|n, base|
      nn = if /old \+ (\d+)/ =~ @op
        n + $1.to_i
      elsif /old \* (\d+)/ =~ @op
        n * $1.to_i
      else
        n * n
      end
      nn %= base if base
      nn
    }

    @test = lines[3].split('divisible by').last.to_i
    @tidx = lines[4].split.last.to_i
    @fidx = lines[5].split.last.to_i
    @test_target = Proc.new {|n| n % @test == 0 ? @tidx : @fidx }

    @total_inspected = 0
  end

  def play(base)
    (0...items.size).map do |idx|
      item = items.shift
      @total_inspected += 1
      nitem = operation.call(item, base) / 3
      [nitem, test_target.call(nitem)]
    end
  end

  def play2(base)
    (0...items.size).map do |idx|
      item = items.shift
      @total_inspected += 1
      nitem = operation.call(item, base)
      [nitem, test_target.call(nitem)]
    end
  end

  def push(item)
    @items.push(item)
  end

  def to_s
    "Monkey #{index}: items: #{items&.join(", ")}"
  end
  alias_method :inspect, :to_s
end

def round(monkeys)
  base = monkeys.map(&:test).inject(:*)
  monkeys.each do |monkey|
    monkey.play(base).each do |nitem, tidx|
      monkeys[tidx].push(nitem)
    end
  end
end

def round2(monkeys)
  base = monkeys.map(&:test).inject(:*)
  monkeys.each do |monkey|
    monkey.play2(base).each do |nitem, tidx|
      monkeys[tidx].push(nitem)
    end
  end
end

def parse_monkey(monkey_lines)
  Monkey.new(monkey_lines)
end

def parse_input(input)
  input.split("\n\n").map {|ml| parse_monkey(ml.split("\n")) }
end

def part1(monkeys)
  20.times do
    round(monkeys)
  end
  monkeys.map(&:total_inspected).sort.reverse.take(2).inject(:*)
end

def part2(monkeys)
  10000.times do |i|
    round2(monkeys)
  end
  monkeys.map(&:total_inspected).sort.reverse.take(2).inject(:*)
end

lines = read_input
p part1 parse_input(lines)
p part2 parse_input(lines)

File: adventofcode/2022/day11/day11.hs
import Data.List (isPrefixOf, sortBy)

split :: (a -> Bool) -> [a] -> [[a]]
split p xs = f xs []
  where f [] agg = [agg]
        f (y : ys) agg
          | p y = agg : f ys []
          | otherwise = f ys (agg ++ [y])

data Monkey = Monkey
  { index :: Int
  , items :: [Int]
  , operation :: (Int -> Int -> Int)
  , testn :: Int
  , tidx :: Int
  , fidx :: Int
  , total_inspected :: Int
  }

instance Show Monkey where
  show (Monkey { index, items, testn, tidx, fidx, total_inspected }) =
    "Monkey: " ++ show index ++ " items: " ++ show items ++
      " testn: " ++ show testn ++ " tidx: " ++ show tidx ++ " fidx: " ++
        show fidx ++ " total_inspected: " ++ show total_inspected ++ "\n"

parseIndex :: String -> Int
parseIndex = read . last . words . head . split (== ':')

parseItems :: String -> [Int]
parseItems = map read . split (== ',') . last . split (== ':')

square :: Int -> Int
square x = x * x

parseOperation :: String -> (Int -> Int -> Int)
parseOperation os
  | op == " old * old" = (mod . square)
  | isPrefixOf " old *" op = (mod . (* opa))
  | otherwise = (mod . (+ opa))
    where op = last $ split (== '=') os
          opa = read $ last $ words op

parseTest :: [String] -> (Int, Int, Int)
parseTest [l0, l1, l2] = (testn, tidx, fidx)
  where tidx = read $ last $ words l1
        fidx = read $ last $ words l2
        testn = read $ last $ words l0

parseMonkey :: [String] -> Monkey
parseMonkey [m0,m1,m2,m3,m4,m5] = Monkey index items operation testn tidx fidx 0
  where index = parseIndex m0
        items = parseItems m1
        operation = parseOperation m2
        (testn, tidx, fidx) = parseTest [m3,m4,m5]

test :: Monkey -> Int -> Int
test (Monkey { testn, tidx, fidx }) i
  | ((== 0) . (`mod` testn)) i = tidx
  | otherwise = fidx

type PlayFN = Monkey -> Int -> [(Int, Int)] -> (Monkey, [(Int, Int)])

play :: PlayFN
play m@(Monkey { items = [] }) _ nt = (m, nt)
play m@(Monkey
  { items = (i:is)
  , operation
  , total_inspected
  }) base nt = play nmonkey base nnt
  where nitem = (div (operation i base) 3)
        nidx = test m nitem
        nnt = nt ++ [(nitem, nidx)]
        nti = total_inspected + 1
        nmonkey = m { items = is, total_inspected = nti }

play2 :: PlayFN
play2 m@(Monkey { items = [] }) _ nt = (m, nt)
play2 m@(Monkey
  { items = (i:is)
  , operation
  , total_inspected
  }) base nt = play2 nmonkey base nnt
  where nitem = operation i base
        nidx = test m nitem
        nnt = nt ++ [(nitem, nidx)]
        nti = total_inspected + 1
        nmonkey = m { items = is, total_inspected = nti }

throwItem :: [Monkey] -> (Int, Int) -> [Monkey]
throwItem ms (nitem, nidx) = (take nidx ms) ++ [nm] ++ (drop (nidx+1) ms)
  where m = ms !! nidx
        nm = m { items = (items m ++ [nitem]) }

throwItems :: [Monkey] -> [(Int, Int)] -> [Monkey]
throwItems ms [] = ms
throwItems ms (n:ns) = throwItems nms ns
  where nms = throwItem ms n

playRound :: PlayFN -> Int -> [Monkey] -> Int -> [Monkey]
playRound playf base ms idx
  | idx == (length ms) = ms
  | otherwise = playRound playf base fms (idx + 1)
  where m = ms !! idx
        (nm, nis) = playf m base []
        nms = throwItems ms nis
        fms = (take idx nms) ++ [nm] ++ (drop (idx+1) nms)

playRounds :: PlayFN -> Int -> Int -> [Monkey] -> [Monkey]
playRounds playf base 0 ms = ms
playRounds playf base n ms = playRounds playf base (n-1) nms
  where nms = playRound playf base ms 0

revSort = sortBy (flip compare)

top2Active :: [Monkey] -> Int
top2Active = product . take 2 . revSort . map total_inspected

part1 :: [Monkey] -> Int
part1 ms = top2Active $ playRounds play base 20 ms
  where base = product $ map testn ms

part2 :: [Monkey] -> Int
part2 ms = top2Active $ playRounds play2 base 10000 ms
  where base = product $ map testn ms

main :: IO ()
main = interact $ unlines . map show . sequence [part1, part2] . map parseMonkey . split (== "") . lines
